articles/110_interface_definition.md

```diff
@@ -0,0 +1,214 @@
+
+
+## Conventions
+
+### Naming of messages and services
+
+Each file contains a single message or service.
+Message files use the extension `.msg`, service files use the extension `.srv`.
+
+Both file names must use an upper camel case name and only consist of alphanumeric characters.
+
+
+### Naming of fields
+
+Field names must be lowercase alphanumeric characters with underscores for separating words.
+They must not start or end with an underscore and never have two consecutive underscores.
+
+
+### Naming of constants
+
+Constant names must be uppercase alphanumeric characters with underscores for separating words.
+They must not start or end with an underscore and never have two consecutive underscores.
+
+
+## Syntax
+
+The message and service definitions are text files.
+
+
+### Comments
+
+The character `#` starts a comment, which terminates at the end of the line on which it occurs.
+
+
+### Message file format
+
+A line can either a field definition or a constant definition.
+
+
+#### Field definition
+
+A field definition has the following structure:
+
+    <type> <name> <optional_default_value>
+
+
+#### Constant definition
+
+A constant definition has the following structure:
+
+    <type> <name>=<value>
+
+
+#### Types
+
+A `<type>` is defined by its *base type* and optional *array specifier*.
+
+The *base type* can be one of the following:
+
+* a primitive type from the list above: e.g. `int32`
+* a string with an upper boundary: `string<=N` to limit the length of the string to `N` characters
+* a complex type referencing another message:
+  * an *absolute* reference of a message: e.g. `some_package/SomeMessage`
+  * a *relative* reference of a message within the same package: e.g. `OtherMessage`
+
+The *array specifier* can be one of the following:
+
+* a static array is described by the suffix `[N]` where `N` is the fixed size of the array
+* an unbounded dynamic array is described by the suffix `[]`
+* a bounded dynamic array is described by the suffix `[<=N]`  where `N` is the maximum size of the array
+
+
+#### Values
+
+Depending on the type the following values are valid:
+
+* `bool`:
+  * `true`, `1`
+  * `false`, `0`
+* `byte`:
+  * an unsigned integer value in the following interval `[0, 255]`
+* `char`:
+  * an integer value in the following interval `[-128, 127]`
+* `float32` and `float64`:
+  * a decimal number using a dot (`.`) as the separator between the integer-part and fractional-part.
+* `int8`, `int16`, `int32` and `int64`:
+  * an integer value in the following interval `[- 2 ^ (N - 1), 2 ^ (N - 1) - 1]` where `N` is the number of bits behind `int`
+* `uint8`, `uint16`, `uint32` and `uint64`:
+  * an unsigned integer value in the following interval `[0, 2 ^ N - 1]` where `N` is the number of bits behind `uint`
+* `string`:
+  * a string value which can optionally be quoted with either single quotes (`'`) or double quotes (`"`)
```

Comment:
Are quotes optional or are they required and the type of the quotes are optional? This reads like the strings could be unquoted.